extensions/2.0/Vendor/KITTYCAD_boundary_representation/README.md

```diff
@@ -0,0 +1,422 @@
+
+#### Base structure
+
+```json
+{
+  "type": "curveType",
+  "curveType": {
+    "curveSpecificValue": 1
+  },
+  "domain": {
+    "min": 0,
+    "max": 1
+  }
+}
+```
+
+#### Line (start & direction)
+
+```json
+"line": {
+  "start": [-1, -1, -1],
+  "direction": [1, 1, 1],
+}
+```
+
+#### Line (start & end)
+
+
+```json
+"line": {
+  "start": [-1, -1, -1],
+  "end": [1, 1, 1],
+}
+```
+
+#### Circle
+
+
+```json
+"circle": {
+  "radius": 1,
+  "origin": [0, 0, 0],
+  "normal": [0, 0, 1],
+  "xbasis": [1, 0, 0]
+}
+```
+
+#### [NURBS](https://en.wikipedia.org/wiki/Non-uniform_rational_B-spline)
+
+
+```json
+"nurbs": {
+  "controlPoints": [
+    [1.0, 0.0, 0.0, 1.0],
+    [1.0, 1.0, 0.0, 0.7071],
+    [0.0, 1.0, 0.0, 1.0]
+  ],
+  "knotVector": [0, 0, 0, 1, 1, 1],
+  "order": 3
+}
+```
+
+## Surface
+
+A _surface_ defines a function that evaluates two parameters to a point in 3D space.
+
+Surfaces are referenced by _faces_ to define the implicit geometry bounded by _loops_.
+
+#### Base structure
+
+```json
+{
+  "type": "surfaceType",
+  "surfaceType": {
+    "surfaceSpecificValue": 1
+  },
+  "domain": {
+    "min": [0, 0],
+    "max": [1, 1]
+  }
+}
+```
+
+#### Plane
+
+```json
+"plane": {
+  "normal": [0, 0, 1],
+  "point": [0, 0, 0],
+  "xbasis": [1, 0, 0]
+}
+```
+
+#### Cylinder (extruded circle)
+
+```json
+"cylinder": {
+  "circle":  {
+    "radius": 1,
+    "origin": [0, 0, 0],
+    "normal": [0, 0, 1],
+    "xbasis": [1, 0, 0]
+  }
+}
+```
+
+#### Torus (revolved circle)
+
+```json
+"torus": {
+  "origin": [0, 0, 0],
+  "radius": 5,
+  "circle":  {
+    "radius": 1,
+    "normal": [0, 0, 1],
+    "xbasis": [1, 0, 0]
+  }
+}
+```
+
+#### Sphere
+
+```json
+"sphere": {
+  "horizon": {
+    "origin": [0, 0, 0],
+    "radius": 1,
+    "normal": [0, 0, 1],
+    "xbasis": [1, 0, 0]
+  }
+}
+```
+
+#### [NURBS](https://en.wikipedia.org/wiki/Non-uniform_rational_B-spline)
+
+```json
+"nurbs": {
+  "controlPoints": [
+    [0, 0, -1, 1]
+    [0, 2, -1, 1],
+    [1, 0, -1, 0.7071],
+    [1, 2, -1, 0.7071],
+    [1, 0, 0, 1],
+    [1, 2, 0, 1],
+  ],
+  "numControlPoints": [3, 2],
+  "numKnots": [6, 4],
+  "knotVector": [3, 3, 3, 4, 4, 4, 0, 0, 2, 2],
+  "order": [3, 2]
+}
+```
+
+## Orientation
+
+Many objects in boundary representation have two _orientation states_. Such objects are called _orientable_ objects. These are often given colloquial terms such as 'right/wrong side', 'up/down', 'in/out', 'forward/backward', et cetera. When referencing orientable objects, it is important to state which orientation of the object is desired. The orientation of a referenced object is described as being 'same-sense' or 'opposite-sense'. This extension utilizes the sign bit of floating point numbers to select the desired orientation. A positive sign selects the 'same-sense' and a negative sign selects the 'opposite-sense'.
+
```

Comment:
In the working group meeting, it was raised that `-0` would likely be parsed as `0` by some JSON parsers. Also, it was raised that the IEEE754 specification says `-0 == 0` should evaluate as true. These are both good points, so I will come up with an alternative.

I have a few ideas:

1. Encode using an object `{ "index": 0, "reverse": true }`. This was actually the original proposal but since this use case is so common, this approach was deemed too verbose/bloated.
2. Use indices starting at 1. This avoids the tricky zeroth index but would be inconsistent with the core specification. I would like to avoid this one if possible.
3. Encode as a string instead, i.e,. `"-0"` instead of `-0`. This could work; however, it was raised this has its own set of problems. For example, glTF JSON permits additional integer representations such as exponential notation.
4. Move this data into the binary blob and encode as a one's complement integer.

Reply:
A couple more ideas:

5. Encode using two integers per array, an index followed by a sign.

```json
{
    "edges": [
        0, 1,
        1, -1,
        2, 1,
        3, -1
    ]
}
```

6. Encode the sign using a separate paired array, which could be omitted if all items are same-sense:

```json
{
    "edges": [0, 1, 2, 3],
    "edgeOrientations": [1, -1, 1, -1]
}
```

Reply:
I've reworked this as an integer/orientation pair. It looks like this:

Before

```json
{
    "loops": [0, -1],
    "surface": 0
}
```

After

```json
{
    "loops": [
        [0, 1],
        [1, -1]
    ],
    "surface": [0, 1]
}
```

`[index, 1]` could be implied by `[index]` but I'll keep it explicit for now.